### ISSUE TITLE
Staging indexer halted: negative number of shares

### ISSUE THREAD

**Issue by csillag** [oasisprotocol/nexus#869]
**This is an old, closed issue, migrated here for historical reference. No action necessary.**

Mainnet staging indexer is currently stuck in Cobalt, height 4743434, for having dead-reckoned a negative number of shares:

```plain
{"analyzer":"consensus","caller":"block.go:250","err":"query 63 &{\n    UPDATE chain.delegations\n      SET shares = shares - $3\n        WHERE delegatee = $1 AND delegator = $2 [oasis1qqekv2ymgzmd8j2s2u7g0hhc7e77e654kvwqtjwm oasis1qryftd0e9588mrd93pznmmuhulu6rzzgngmtkl8x 1000000000000]}: ERROR: value for domain uint_numeric violates check constraint \"uint_numeric_check\" (SQLSTATE 23514)","height":4743434,"level":"error","module":"analysis_service","msg":"error processing block","slow_sync":true,"ts":"2023-09-18T05:56:46.565278975Z"}
```

  

Investigate, fix. Might be interesting to compare the dead-reckoned shares with the node-reported shares at that height.

  

Production

```plain
UPDATE chain.delegations\n      SET shares = shares - $3\n        WHERE delegatee = $1 AND delegator = $2 [oasis1qqekv2ymgzmd8j2s2u7g0hhc7e77e654kvwqtjwm oasis1qryftd0e9588mrd93pznmmuhulu6rzzgngmtkl8x 1000000000000]}: ERROR: value for domain uint_numeric violates check constraint \"uint_numeric_check\" (SQLSTATE 23514)
```

## Discussion:

Might be worth creating a little bisection tool for tracking down problems with our dead reckoning. We can find the height at which our dead-reckoned value and the node's value started diverging.

The tool would take as inputs:
- A min and max height between which to bisect, duh.
- A function that recreates the bisected value at a given height h. We don't have actual old dead-reckoned values readily stored in the DB, but we do store all events in the DB, so we can recreate the dead-reckoned value with a sufficiently fancy SUM().
- A function that fetches the expected value from the node.

This approach should be helpful for this bug, and also in the future if one of the negative-balance bugs comes back to haunt us.

**Comment by csillag** [comment]
Further comments by @mitjat 

TLDR: We don't support dead-reckoning AddEscrow in Damask. So let's avoid it by using fast sync.

Full writeup:

I wrote a bisection tool (https://github.com/oasisprotocol/nexus/pull/564) and a small tweak to guard against the same problem in the future (https://github.com/oasisprotocol/nexus/pull/565).

Bisection finds the following discrepancy between db and node for the perceived number of shares (for the known problematic delegator-delegatee pair):

```
*GOOD height 4687618:   node 0                 db 0              

 BAD  height 4687619:   node 111589494653242   db 0 
```

At that height, there's an `AddEscrow` event without a `new_shares` field. Nexus uses that field to dead-reckon the delegator's amount of shares; since the field is absent, we cannot dead-reckon correctly.

We already "knew" that such AddEscrow events might appear: I [added this comment](https://github.com/oasisprotocol/nexus/blob/9baa9f663ceaa30b10960f1220a09855cec1e334/coreapi/v21.1.1/staking/api/api.go#L86-L86) about Cobalt but didn't realize the implications at the time.



*I'll go ahead with what I think the most expedient solution: Use fast-sync up to a recent height* (where `AddEscrow` always contains `new_shares`). This way, these old faulty AddEscrow events won't be used for dead-reckoning, and escrow balances will be fetched via `StateToGenesis`.

A more thorough solution would be to separately query the node for the absolute number of shares whenever we encounter an `AddEscrow` without `new_shares`. But that's realistically not needed.

NOTE: This is `prod mainnet`; it has only used slow-sync so far. OTOH `staging mainnet` is already using fast-sync, but it's been unable to use `StateToGenesis` to transition into slow-sync because the `StateToGenesis` RPC times out at some proxy. We'll see if `prod mainnet` hits the same issue. But even if it does, we'll at worst move from one kind of stuck (negative number of balances, analyzer can't proceed) to another kind of stuck (`StateToGenesis` can't run).


NOTE 2: I didn't quite need bisection to find the problem. There are only 3 events involved with the delegator in the entire set of scanned heights: 2x `staking.transfer`, 1x `staking.escrow.add`

But I got nerd-sniped into finishing the bisection tool; I do think there's a good chance it will come in handy again, and for less trivial cases.

**Comment by csillag** [comment]
Last update by @mitjat 

> `prod mainnet` nexus is currently catching up via fast-sync; ETA Monday, November 20.
